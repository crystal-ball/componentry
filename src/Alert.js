import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';

import { Button, Icon } from '../index';
import { visibilityTransitionLength } from './utils/configurations';

/**
 * Alerts provide contextual feedback.
 * To Document:
 * - Internal fallback _hidden method, onDismiss preferred
 * - Internal fade and hidden states
 * - onDismiss and dismissable configurations
 * - alerts only have info classes, not primary or secondary, b/c they are intended
 *   to be used as an actual alert with context, denoted by role=alert. For non-alert
 *   information blocks a card with primary or secondary colors can be used.
 */
export default class Alert extends Component {

  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    color: PropTypes.string,
    dismissable: PropTypes.bool,
    onDismiss: PropTypes.func,
    visibilityTransitionLength: PropTypes.number
  }

  static contextTypes = {
    visibilityTransitionLength: PropTypes.number
  }

  static defaultProps = {
    children: null,
    className: '',
    color: '',
    dismissable: true,
    onDismiss: null,
    visibilityTransitionLength: null
  }

  // Fade controls visibility status and hidden controls DOM position status
  state = {
    fade: false,
    hidden: false
  }

  /**
   * Backup onDismiss for dismissable alerts without a passed onDismiss. Note that
   * this is just a convenience method. Passing an `onDismiss` that handles updating
   * application state to dismiss an alert is preferred.
   */
  _dismiss = () => {
    // props has precedence to allow for single instance overrides, context can be
    // used for app wide configs, fall back to defaults
    const timer = this.props.visibilityTransitionLength
      || this.context.visibilityTransitionLength
      || visibilityTransitionLength;

    // Will immediately set BS 'fade' class to transition opacity to 0
    this.setState({ fade: true });
    // Roughly when transition is finished, add aria-hidden to element to remove display
    setTimeout(() => {
      this.setState({ hidden: true });
    }, timer);
  }

  // Render
  // ---------------------------------------------------------------------------
  /**
   * If the alert is dismissable create a close button, otherwise return null
   * @return {Component|null}
   */
  renderClose = () => {
    let { dismissable, onDismiss } = this.props;
    // If alert is dismissable, but an onDismiss wasn't passed, use our internal
    // fade out
    onDismiss = onDismiss || this._dismiss;

    return dismissable ?
      (<Button link onClick={onDismiss} className='close' aria-label='close'>
        <Icon icon='close' />
      </Button>)
      : null;
  }

  render() {
    // declare dismissable, onDismiss to keep ...other spread render safe
    let {
      children,
      className='',
      color='',
      dismissable, // eslint-disable-line
      onDismiss, // eslint-disable-line
      visibilityTransitionLength, // eslint-disable-line
      ...other
    } = this.props;
    let { fade, hidden } = this.state;
    let ariaHidden = hidden ? 'true' : 'false';
    let _className = classnames('alert', className, {
      [`alert-${color}`]: color,
      fade: fade
    });

    let close = this.renderClose();

    return (
      <div className={_className} role='alert' {...other} aria-hidden={ariaHidden}>
        <div className='alert-content'>
          {children}
        </div>
        {close}
      </div>
    );
  }
}
