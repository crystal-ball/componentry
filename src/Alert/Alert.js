import React from 'react'
import element from '../element'
import { closeBase } from '../Close/Close'
import { useTheme } from '../Theme/Theme'
import staticComponent from '../factories/static-component'
import { useActive, useVisible } from '../hooks'

/**
 * @typedef {Object} Props
 * @property {string} Props.ariaTitle Set a specific aria title
 * @property {string} Props.color Set the theme color of the alert
 * @property {boolean} Props.dismissible If true the alert is dismissible
 * @property {boolean} Props.outline If true, the alert will have outline styles
 */

/**
 * [Alert component üìù](https://componentry.design/components/alert)
 */
export default function Alert(props) {
  /** @type {Props} */
  const {
    children,
    active: propsActive,
    ariaTitle,
    color,
    deactivate,
    dismissible,
    outline,
    ...rest
  } = { ...useTheme('Alert'), ...useActive(), ...props }

  const { active, visible } = useVisible(propsActive)

  return element({
    'role': 'alert',
    'componentCx': {
      'alert': true,
      'fade': dismissible, // Only include opacity transition class for disimissible alerts
      visible,
      [`alert-${color}`]: color,
      'alert-outline': outline,
    },
    // ‚ö†Ô∏è Only include aria-hidden value if the alert is dismissible
    'aria-hidden': dismissible ? String(!active) : undefined,
    'children': (
      <>
        {/* Provide the alert color context for screen readers */}
        <div className='sr-only'>{ariaTitle || `${color} alert`}</div>

        {/* Alert contents */}
        <div className='alert-content'>{children}</div>

        {/* Render a close button or null depending on configs */}
        {dismissible && (
          <Alert.Close
            className={`alert-close font-color-${color}`}
            onClick={deactivate}
          />
        )}
      </>
    ),
    ...rest,
  })
}
Alert.displayName = 'Alert'

/**
 * [Alert close component üìù](https://componentry.design/components/alert)
 */
Alert.Close = staticComponent('AlertClose', closeBase)
