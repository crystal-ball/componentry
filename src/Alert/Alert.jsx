// @flow
import React, { createElement } from 'react'
import { object, shape } from 'prop-types'
import classNames from 'classnames'

import Close from '../Close/Close'
import withActive from '../withActive/withActive'

import type { ThemeColors } from '../utils/theme'
import type { ActiveProps } from '../withActive/withActive'
import type { ElementProps } from '../component-factories/element-factory'

type Props = {
  /**
   * Provides contextual information to screen readers that sighted users would gain
   * from the alert color. Defaults to: 'COLOR alert', eg 'Success alert'
   */
  ariaTitle: string,
  /**
   * Theme color for alert
   */
  color: ThemeColors,
  /**
   * Pass `false` to suppress alert dismiss button
   */
  dismissible?: boolean,
  /**
   * Length of opacity transition, defaults to 300ms or `THEME` value if set using
   * `ThemeProvider`.
   */
  transitionDuration?: number,
} & ActiveProps &
  ElementProps

type Context = { [string]: { [string]: any } }

/**
 * Alerts provide contextual feedback to users. Alerts are available in the info
 * theme colors success, info, warning and danger. They are not available in primary
 * or secondary theme colors because they are intended to be used for alerting with
 * context. For non alert information blocks a card with theme color primary or
 * secondary can be used.
 */
const Alert = (props: Props, { THEME = {} }: Context) => {
  const componentCtx = THEME.Alert || {}
  const {
    ariaTitle,
    activate, // prevent dom inclusion
    active,
    as,
    children,
    className,
    color,
    dismissible,
    deactivate,
    visible,
    ...rest
  } = { ...componentCtx, ...props }

  return createElement(
    as || 'div',
    {
      role: 'alert',
      className: classNames('alert', 'fade', componentCtx.className, className, {
        [`alert-${color}`]: color,
        show: visible,
      }),
      // hidden state is updated after active opacity transition
      'aria-hidden': active ? 'false' : 'true',
      ...rest,
    },
    // Provide the alert color context for screen readers
    <div className="sr-only">{ariaTitle || `${color} alert`}</div>,
    // Alert contents:
    <div className="alert-content">{children}</div>,
    // Render a close button or null depending on configs
    dismissible && <Close onClick={deactivate} className={`text-${color}`} />,
  )
}

Alert.displayName = 'Alert'
Alert.defaultProps = {
  /* eslint-disable react/default-props-match-prop-types */
  active: true,
  visible: true,
  dismissible: false,
}
Alert.contextTypes = { THEME: shape({ Alert: object }) }

const withActiveAlert = withActive({ defaultActive: true, transitionState: true })(
  Alert,
)

export default withActiveAlert
