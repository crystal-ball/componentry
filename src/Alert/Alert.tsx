/* eslint-disable @typescript-eslint/no-empty-interface */

import React from 'react'
import { closeBase } from '../Close/Close'
import { useTheme } from '../Theme/Theme'
import { useActive, useVisible } from '../hooks'
import { ComponentBaseProps } from '../utils/types'
import { element } from '../utils/element-creator'
import { staticComponent } from '../utils/static-component-builder'

interface CardCloseProps extends ComponentBaseProps<'button'> {}

interface AlertProps extends ComponentBaseProps<'div'> {
  /** Sets a custom aria title */
  ariaTitle?: string
  /** Sets the theme color of the alert */
  color?: 'primary' | 'success' | 'warning' | 'critical'
  /** Deactivate handler called on click of Alert dismiss button */
  deactivate?: (event: React.MouseEvent<HTMLButtonElement>) => void
  /** Toggles alert dismissible feature */
  dismissible?: boolean
  /** Toggles alert outline styles */
  variant?: 'primary' | 'outline'
}

interface Alert {
  (props: AlertProps): React.ReactElement
  displayName: 'Alert'
  /**
   * [Alert close component üìù](https://componentry.design/components/alert)
   */
  Close: React.FC<CardCloseProps>
}

/**
 * [Alert component üìù](https://componentry.design/components/alert)
 */
export const Alert: Alert = (props) => {
  const {
    children,
    active: propsActive,
    ariaTitle,
    color,
    deactivate,
    dismissible,
    variant = 'primary',
    ...rest
  } = { ...useTheme<AlertProps>('Alert'), ...useActive(), ...props }

  const { active, visible } = useVisible(propsActive)

  return element('Alert', {
    'role': 'alert',
    'componentCx': [
      `alert-${variant}`,
      {
        [`alert-${color}`]: color,
        fade: dismissible, // Only include opacity transition class for disimissible alerts
        visible,
      },
    ],
    'aria-hidden': dismissible ? String(!active) : 'false',
    'children': (
      <>
        {/* Provide the alert color context for screen readers */}
        <div className='sr-only'>{ariaTitle || `${color} alert`}</div>

        {/* Alert contents */}
        <div className='alert-content'>{children}</div>

        {/* Render a close button or null depending on configs */}
        {dismissible && (
          <Alert.Close
            className={`alert-close font-color-${color}`}
            onClick={deactivate}
          />
        )}
      </>
    ),
    ...rest,
  })
}
Alert.displayName = 'Alert'

Alert.Close = staticComponent<CardCloseProps>('AlertClose', closeBase)
