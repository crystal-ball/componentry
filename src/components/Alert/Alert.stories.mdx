import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs'
import { BADGE } from '@geometricpanda/storybook-addon-badges'

import { Active, Alert, Link, Text } from '../..'

<Meta
  title='Components/Alert'
  component={Alert}
  parameters={{ badges: [BADGE.EXPERIMENTAL] }}
/>

# Alert

Alerts can contain any HTML elements, and can be dismissible or static. A dismissible
Alert requires Componentry Active `active` and `deactivate` props, which can be passed
directly or the Alert can be nested inside of an Active component to automatically create
an uncontrolled Alert.

<Canvas>
  <Story name='Alert'>
    <Alert color='primary'>
      <Text variant='h3' className='ðŸ…²AlertHeading'>
        Well done!
      </Text>
      <Text>You successfully read this important alert message.</Text>
      <hr className='my-2' />
      <Link href='#demo' className='ðŸ…²AlertLink'>
        Go home
      </Link>
    </Alert>
  </Story>
</Canvas>

Styles are also included for: `alert-heading`, `alert-link`, and `<hr />` elements.

## Dismissible

Pass `dismissible` to an Alert to show a close icon which will call the `deactivate` prop
on click.

<Alert color='success' dismissible>
  <Text variant='h3' className='ðŸ…²AlertHeading'>
    Success!
  </Text>
  <Text>You created a dismissible alert ðŸ‘‹</Text>
</Alert>

## Accessibility

### Aria title

Provide an aria title for alerts with better accessibility.

<Alert color='primary' ariaTitle='Alert provides context on setting the aria title'>
  Pass an `ariaTitle` to provide more context.
</Alert>
