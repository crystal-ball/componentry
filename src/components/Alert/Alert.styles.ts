import { getMergedConfig } from '../../plugin-postcss/configs'

const { theme } = getMergedConfig()

//                                      Componentry <Alert /> styles
// -----------------------------------------------------------------

export const alertStyles: Record<string, unknown> = {
  // BASE
  '.ðŸ…²Alert-base': {
    // Make the alert container a flex container by default with space-between
    // to align close button to right side
    display: 'flex',
    alignItems: 'flex-start',
    justifyContent: 'space-between',
  },

  '.ðŸ…²Alert-dismissible': {
    opacity: 1,
    transition: 'opacity 0.2s cubic-bezier(0.4, 0, 0.2, 1)',
  },
  '.ðŸ…²Alert-dismissed': {
    opacity: 0,
  },

  // VARIANTS
  '.ðŸ…²Alert-filled': {
    padding: theme.spacing[2],
    background: theme.colors.primary[100],
    color: theme.colors.primary[500],
    border: '1px solid transparent',
    borderRadius: theme.borderRadius.md,
    borderColor: theme.colors.primary[300],

    '& .ðŸ…²AlertLink': {
      fontWeight: theme.fontWeight.bold,
    },

    // TODO: Recreate this with a Divider component
    // '& hr': { borderTopColor: theme.colors.primary[300] },
  },

  // ELEMENTS
  '.ðŸ…²AlertContent': {
    flexGrow: 1,
    flexShrink: 1,
  },

  '.ðŸ…²AlertHeading': {
    color: 'inherit',
    marginBottom: theme.spacing[2],
  },

  '.ðŸ…²AlertClose': {
    flexShrink: 0,
    marginLeft: theme.spacing[1],
    color: 'inherit',
  },
}

// --------------------------------------------------------
// COLORS

// Example of adding theme colors to an alert variant
// const themeColors = ['info', 'success', 'warning', 'error'] as const
// themeColors.forEach((color) => {
//   Alert[`.ðŸ…²Alert-filled.ðŸ…²Alert-${color}Color`] = {
//     'background': theme.colors[color][100],
//     'borderColor': theme.colors[color][300],
//     'color': theme.colors[color][500],

//     '& .ðŸ…²Alert-link': {
//       color: theme.colors[color][500],
//       fontWeight: theme.fontWeight.bold,
//     },

//     '& hr': {
//       borderTopColor: theme.colors[color][300],
//     },
//   }
// })
