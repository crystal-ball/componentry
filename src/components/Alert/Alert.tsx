import React from 'react'
import { useActive, useVisible } from '../../hooks'
import { createElement } from '../../utils/create-element'
import { createStaticComponent } from '../../utils/create-static-component'
import { MergeTypes, Resolve } from '../../utils/types'
import { UtilityProps } from '../../utils/utility-props'
import { closeBase } from '../Close/Close'
import { useThemeProps } from '../Provider/Provider'

/** Module augmentation interface for overriding component props' types */
export interface AlertPropsOverrides {}

export interface AlertPropsBase {
  /** Sets a custom aria title */
  ariaTitle?: string
  /** Sets the theme color of the alert */
  color?: 'primary' | 'success' | 'warning' | 'critical'
  /** Deactivate handler called on click of Alert dismiss button */
  deactivate?: (event: KeyboardEvent | MouseEvent | TouchEvent | React.MouseEvent) => void
  /** Toggles alert dismissible feature */
  dismissible?: boolean
  /**
   * Controls when the component content is mounted where:
   * - `'always'` - The content will be mounted when the element is both visible
   *   and not visible
   * - `'visible'` - The content will only be mounted when visible, when not
   *   visible the contents are not rendered for performance.
   */
  mounted?: 'always' | 'visible'
  /** Sets the alert variant style */
  variant?: 'filled'
}

export type AlertProps = Resolve<MergeTypes<AlertPropsBase, AlertPropsOverrides>> &
  Omit<UtilityProps, 'color'> &
  React.ComponentPropsWithoutRef<'div'>

export interface AlertCloseProps
  extends UtilityProps,
    React.ComponentPropsWithoutRef<'button'> {}

export interface Alert {
  (props: AlertProps): React.ReactElement | null
  displayName: 'Alert'
  /**
   * [Alert close component üìù](https://componentry.design/components/alert)
   */
  Close: React.FC<AlertCloseProps>
}

/**
 * [Alert component üìù](https://componentry.design/components/alert)
 * @alpha
 */
export const Alert: Alert = (props) => {
  const {
    children,
    active: _active,
    ariaTitle,
    color,
    deactivate,
    mounted = 'visible',
    dismissible,
    variant = 'filled',
    ...rest
  } = { ...useThemeProps('Alert'), ...useActive(), ...props }

  const { active, visible } = useVisible(_active)

  if (dismissible && !active && mounted === 'visible') return null

  return createElement({
    role: 'alert',
    componentClassName: [
      `C9Y-Alert-base C9Y-Alert-${variant}`,
      {
        [`C9Y-Alert-${color}Color`]: color,
        'C9Y-Alert-dismissible': dismissible,
        'C9Y-Alert-dismissed': dismissible && !visible,
      },
    ],
    'aria-hidden': dismissible ? String(!active) : 'false',
    children: (
      <>
        {/* Provide the alert color context for screen readers */}
        <div className='sr-only'>{ariaTitle || `${color} alert`}</div>

        {/* Alert contents */}
        <div className='C9Y-AlertContent'>{children}</div>

        {/* Render a close button or null depending on configs */}
        {dismissible && <Alert.Close className='C9Y-AlertClose' onClick={deactivate} />}
      </>
    ),
    ...rest,
  })
}
Alert.displayName = 'Alert'

Alert.Close = createStaticComponent<AlertCloseProps>('AlertClose', closeBase)
