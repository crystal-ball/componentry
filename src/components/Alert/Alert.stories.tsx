import type { Meta, StoryObj } from '@storybook/react'

import { Active } from '../Active/Active'
import { Link } from '../Link/Link'
import { Text } from '../Text/Text'
import { Alert } from './Alert'

const meta: Meta<typeof Alert> = {
  component: Alert,
  parameters: {
    docs: {
      description: {
        component:
          'Alerts can contain any HTML elements, and can be dismissible or static. A dismissible Alert requires Componentry Active `active` and `deactivate` props, which can be passed directly or the Alert can be nested inside of an Active component to automatically create an uncontrolled Alert.',
      },
    },
  },
}

export default meta
type Story = StoryObj<typeof Alert>

export const Primary: Story = {
  args: {
    color: 'primary',
    children: (
      <>
        <Text variant='h3' className='C9Y-AlertHeading'>
          Well done!
        </Text>
        <Text>You successfully read this important alert message.</Text>
        <hr className='my-2' />
        <Link href='#demo' className='C9Y-AlertLink'>
          Go home
        </Link>
      </>
    ),
  },
}

export const Dismissible: Story = {
  render: () => (
    <Active defaultActive>
      <Alert color='primary' dismissible>
        <Text variant='h3' className='C9Y-AlertHeading'>
          Success!
        </Text>
        <Text>You created a dismissible alert.</Text>
      </Alert>
    </Active>
  ),
}
