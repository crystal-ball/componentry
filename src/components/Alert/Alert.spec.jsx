import { render, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { describe, it, expect } from 'vitest'

import { elementTests } from '../../test/element-tests'
import { Alert } from './Alert'

describe('<Alert/>', () => {
  // Basic library element test suite
  elementTests(Alert)
  elementTests(Alert.Close)

  it('should render an alert without a close button by default', () => {
    render(<Alert color='success'>Warning!</Alert>)

    expect(screen.getByRole('alert')).toHaveClass(
      'C9Y-Alert-base C9Y-Alert-filled C9Y-Alert-successColor',
    )
    expect(screen.getByRole('alert')).toHaveAttribute('role', 'alert')
  })

  it('should not render a close button if not dismissible', () => {
    render(<Alert color='danger'>Warning!</Alert>)

    expect(screen.queryByLabelText('close')).not.toBeInTheDocument()
  })

  it('should bind passed deactivate to close button', async () => {
    const deactivate = vi.fn()
    render(
      <Alert color='danger' deactivate={deactivate} active dismissible>
        Warning!
      </Alert>,
    )

    expect(screen.getByText('Warning!')).toBeInTheDocument()

    await userEvent.click(screen.getByLabelText('close'))

    // eslint-disable-next-line jest/prefer-called-with
    expect(deactivate).toHaveBeenCalled()

    // Alert visibility state change handler has been overridden, other than calling
    // passed deactivate Alert should still be visible & unchanged
    expect(screen.getByRole('alert')).toHaveAttribute('aria-hidden', 'false')
  })

  // TODO: test active and deactivate handling
})

describe('<Alert /> snapshots', () => {
  it('renders correctly', () => {
    render(
      <Alert color='danger' deactivate={vi.fn()} active dismissible>
        Warning!
      </Alert>,
    )

    expect(screen.getByRole('alert')).toMatchSnapshot()
  })
})
