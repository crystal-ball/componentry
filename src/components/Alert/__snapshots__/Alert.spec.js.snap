// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> snapshots renders correctly 1`] = `
<div
  aria-hidden="false"
  class="ðŸ…²Alert-base ðŸ…²Alert-filled ðŸ…²Alert-dangerColor fade"
  role="alert"
>
  <div
    class="sr-only"
  >
    danger alert
  </div>
  <div
    class="ðŸ…²AlertContent"
  >
    Warning!
  </div>
  <button
    aria-label="close"
    class="ðŸ…²Close-base ðŸ…²AlertClose"
    type="button"
  >
    <svg
      class="ðŸ…²Icon-base ðŸ…²Icon-font icon-close"
      role="img"
    >
      <use
        href="#close"
        xlink:href="#close"
      />
    </svg>
  </button>
</div>
`;
